### PR TITLE
ARROW-2657: [C++] Undo export pthread_once related symbols.

### PR DESCRIPTION
**Pull request by robertnishihara**
This seems to fix https://issues.apache.org/jira/browse/ARROW-2657, but I don't fully understand this code so it may introduce other issues.

This code was introduced in #1953.

cc @wesm @pitrou

### CODE REVIEW

**Comment by pitrou** [comment]
> I don't fully understand this code so it may introduce other issues.

As you can see on Travis-CI (look at the manylinux1 build).

**Comment by pitrou** [comment]
If, as @wesm suggests, "TensorFlow is not respecting the manylinux1 standard and is using newer compilers", I'm not sure how we can reconcile TensorFlow and Arrow :-/

**Comment by pitrou** [comment]
The original problem is that `libarrow.so` and `libplasma.so` both link with libstdc++ and, when the RH devtoolset is used (i.e. when building wheels), some parts of the libstdc++ are linked statically (the parts that aren't available at runtime with an old libstdc++). Then, when loading `libplasma.so` which also loads `libarrow.so`, those static symbols end up instantiated twice in the process (because they are made library-local by the visibility script).

The fix for that was to ensure that those symbols stay process-global, so they that all resolve to a single location.

**Comment by wesm** [comment]
This is a bit of a mess -- I'm reading through the original comments to understand what's happening better. Would this not be a problem in any setting where devtoolset was used to build multiple .so files?

**Comment by wesm** [comment]
Where is the `call_once` call happening? I searched in the codebase and didn't find that anywhere

**Comment by pitrou** [comment]
`call_once` is apparently used under the hood for lazily-initialized static singletons (i.e. when a `static` object is defined inside a function).

**Comment by pitrou** [comment]
> Would this not be a problem in any setting where devtoolset was used to build multiple .so files?

I think it's ok as long as you don't try to hide the "once" symbols using a visibility script.

**Comment by wesm** [comment]
@xhochy is there a problem if we allow these pthread symbols to be visible in the .so? We might have to set up some kind of exclusions file for the visibility check script

**Comment by xhochy** [comment]
We actually hide these symbols to be compatible with other libs. In this case it sadly produces the opposite, so it's fine to exclude them.

**Comment by robertnishihara** [comment]
@xhochy, where does `check_arrow_visibility.sh` live?

**Comment by wesm** [comment]
That's the one, let me know if you need help. I think that may need to be turned into a Python script that can read from an exclusions file

File: python/manylinux1/scripts/check_arrow_visibility.py [hunk 1]
```diff
@@ -16,14 +17,22 @@
...
-nm -D -C /arrow-dist/lib64/libarrow.so > nm_arrow.log
-grep ' T ' nm_arrow.log | grep -v arrow > visible_symbols.log
+import subprocess
 
-if [[ `cat visible_symbols.log | wc -l` -eq 2 ]]
-then
-    exit 0
-fi
+exceptions = [
+    '__once_proxy',
```
Comment (robertnishihara) [inline python/manylinux1/scripts/check_arrow_visibility.py]: If we remove this line, the test fails in Travis with

```
Step 9/14 : RUN /check_arrow_visibility.py
 ---> Running in 3dc1e6cc807c
Traceback (most recent call last):
  File "/check_arrow_visibility.py", line 36, in ?
    raise Exception(symbols)
Exception: [u'_fini', u'_init', u'__once_proxy']
The command '/bin/sh -c /check_arrow_visibility.py' returned a non-zero code: 1
```

**Comment by wesm** [comment]
Is `-pthread` being passed when linking libplasma? 

This gcc thread suggests that may fix the problem https://gcc.gnu.org/bugzilla/show_bug.cgi?id=58104. Try adding `${PTHREAD_LIBRARY}` to https://github.com/apache/arrow/blob/master/cpp/src/plasma/CMakeLists.txt#L66?

**Comment by wesm** [comment]
Although I'm not sure why that particular test would fail =( I can try to tinker with this tomorrow possibly, I don't know where to turn for help

**Comment by wesm** [comment]
Another thing we could try is to not use the global thread pool in Plasma, in case there is some issue with using the global thread pool in `libarrow` in `libplasma` (and we can resolve the shared-library-crossing issue at some later time)

**Comment by wesm** [comment]
We could also try replacing the global thread pool initialization with an explicit use of `std::call_once` in case the issue relates to something having to do with gcc-4.8's implementation of threadsafe static initialization

**Comment by wesm** [comment]
I also think we should get the TensorFlow team involved in this. They bear some responsibility for their choice of a non-conforming compiler toolchain when building their wheels

**Comment by wesm** [comment]
Rebased. Taking a look at this; I am curious if replacing the atomic initialization of the threadpool with a mutex will make the problem go away

**Comment by wesm** [comment]
Looks like PyTorch and TensorFlow trampled each other over this issue also https://github.com/pytorch/pytorch/issues/3059#issuecomment-335668027. Digging some more into this

**Comment by wesm** [comment]
Been banging my head against this for a few hours now. I've been looking at PyTorch's equivalent symbols.map file

https://github.com/pytorch/pytorch/blob/master/tools/pytorch.version

I tried to make it look more like theirs:

```
{
    global:
        extern "C++" {
          *arrow::*;
        };

    local:
        *;
  };
```

But then this causes errors like

```
$ python -c "import pyarrow"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/wesm/code/arrow/python/pyarrow/__init__.py", line 47, in <module>
    from pyarrow.lib import cpu_count, set_cpu_count
ImportError: /home/wesm/local/lib/libarrow.so.10: undefined symbol: _ZNK6google8protobuf7Message11GetTypeNameEv
```

Intuitively, all of the symbols except for the `arrow::` symbols (and other C++-related symbols like typeinfo, etc.) should be hidden. I'm going to have to take a break from this for a day or two before I'll be able to work on it again as it's too frustrating

**Comment by robertnishihara** [comment]
@wesm thanks for looking into this. I'm out of ideas at the moment about the "right" way to solve this. If we can't figure something out then it will likely be possible to workaround by not using the global thread pool in plasma.

**Comment by wesm** [comment]
After looking at this, I'm not sure that will help; we would have to find some other way to have a global thread pool without having a static global in libarrow.so. We should try to better understand the conflict with TensorFlow -- the fact that PyTorch was able to solve the problem by hiding the symbol is promising.

I'm concerned that devtoolset-2 (which is based on gcc 4.8.x) is on the very edge of C++11 support and so there may be something idiosyncratic that's going wrong. I wonder how much trouble we'd be getting ourselves into if we built our wheels with devtoolset-3 (gcc 4.9.x-based)

**Comment by pitrou** [comment]
Do we already have other static singletons in Arrow?

I suppose it could be possible to use our own mutex-backed implementation if that helps (I don't think the `GetCpuThreadPool` performance is that critical - if you're planning to offload tasks to a thread pool, you're obviously ok with enduring the cost of thread synchronization). But I'm curious why TensorFlow is using a different toolchain - obviously that should conflict with the manylinux1 requirements?

**Comment by wesm** [comment]
I tried using mutexes to sychronize -- the issue is having any global variable that is a C++ object it seems -- this causes the compiler to emit the `std::once_call` instructions resulting in the pthread symbols being statically linked. I don't yet understand why these symbols are a conflict. From the PyTorch solution, it seems that getting our symbol exports right should solve the problem

**Comment by pitrou** [comment]
Answering to myself: yes, we already have static singletons in the Arrow codebase. For example in the type factories in `type.cc`.

**Comment by wesm** [comment]
You're right; I'm pretty mystified why this problem only started coming up now

TensorFlow needs a newer version of clang for all the LLVM-stuff they are using I guess

**Comment by pitrou** [comment]
> I tried using mutexes to sychronize

Which idiom did you try exactly?

I'm thinking something like:

```c++

static std::mutex cputhreadpool_mutex;
static CPUThreadPool* cputhreadpool = nullptr;

GetCPUThreadPool() {
  lock_guard lock(cputhreadpool_mutex);
  if (cputhreadpool == nullptr) {
    cputhreadpool = ...;
  }
  return cputhreadpool;
}
```

By putting all `static` variables outside of the function we would ideally remove the `call_once` usage.

**Comment by wesm** [comment]
That's what I tried. I suspect the call_once symbols are getting pulled in from something else (maybe from the condition_variable use?) -- we may be looking at a red herring.

**Comment by pitrou** [comment]
Disassembling the `libarrow.so` that's part of the generated wheel would probably tell us where `call_once` is used exactly.

**Comment by wesm** [comment]
I was just looking at https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables. I will have another go at trying to fix this, but I think we might consider following Google's advice and not having any static globals with non-trivial destructors (i.e. shared_ptr has a non-trivial dtor). I suspect that's what caused these symbols to bubble up in the first place

**Comment by robertnishihara** [comment]
Thanks @wesm! Let us know what you find. I think following that style guide point makes a lot of sense.

**Comment by wesm** [comment]
The main problem with that is that it would probably require explicit initialization of the thread pool, and applications would need to shut down the thread pool explicitly to avoid valgrind warnings. Will let you know what I find

**Comment by wesm** [comment]
I'm getting closer on this. So I'm able to load the various libraries and then load tensorflow:

```
$ python
Python 3.6.5 | packaged by conda-forge | (default, Apr  6 2018, 13:39:56) 
[GCC 4.8.2 20140120 (Red Hat 4.8.2-15)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ctypes
>>> ctypes.CDLL('/home/wesm/miniconda/envs/wheel-test-2/lib/python3.6/site-packages/pyarrow/libarrow.so')
<CDLL '/home/wesm/miniconda/envs/wheel-test-2/lib/python3.6/site-packages/pyarrow/libarrow.so', handle 1948b50 at 0x7fa407a9ff98>
>>> ctypes.CDLL('/home/wesm/miniconda/envs/wheel-test-2/lib/python3.6/site-packages/pyarrow/libarrow_python.so')
<CDLL '/home/wesm/miniconda/envs/wheel-test-2/lib/python3.6/site-packages/pyarrow/libarrow_python.so', handle 196a9a0 at 0x7fa407a9ffd0>
>>> ctypes.CDLL('/home/wesm/miniconda/envs/wheel-test-2/lib/python3.6/site-packages/pyarrow/libarrow_boost_regex.so.1.65.1')
<CDLL '/home/wesm/miniconda/envs/wheel-test-2/lib/python3.6/site-packages/pyarrow/libarrow_boost_regex.so.1.65.1', handle 196c8d0 at 0x7fa407a9ff98>
>>> ctypes.CDLL('/home/wesm/miniconda/envs/wheel-test-2/lib/python3.6/site-packages/pyarrow/libarrow_boost_filesystem.so.1.65.1')
<CDLL '/home/wesm/miniconda/envs/wheel-test-2/lib/python3.6/site-packages/pyarrow/libarrow_boost_filesystem.so.1.65.1', handle 19526b0 at 0x7fa407a9ffd0>
>>> ctypes.CDLL('/home/wesm/miniconda/envs/wheel-test-2/lib/python3.6/site-packages/pyarrow/libarrow_boost_system.so')
<CDLL '/home/wesm/miniconda/envs/wheel-test-2/lib/python3.6/site-packages/pyarrow/libarrow_boost_system.so', handle 196d070 at 0x7fa407a9ff98>
>>> ctypes.CDLL('/home/wesm/miniconda/envs/wheel-test-2/lib/python3.6/site-packages/pyarrow/libparquet.so')
<CDLL '/home/wesm/miniconda/envs/wheel-test-2/lib/python3.6/site-packages/pyarrow/libparquet.so', handle 196d920 at 0x7fa407a9ffd0>
>>> import tensorflow
```

But

```
$ python
Python 3.6.5 | packaged by conda-forge | (default, Apr  6 2018, 13:39:56) 
[GCC 4.8.2 20140120 (Red Hat 4.8.2-15)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyarrow
>>> import tensorflow
Segmentation fault (core dumped)
```

You can see from gdb that importing pyarrow fires up the thread pool:

```
>>> import pyarrow
im[New Thread 0x7fffecc4d700 (LWP 7962)]
[New Thread 0x7fffec44c700 (LWP 7963)]
[New Thread 0x7fffe7c4b700 (LWP 7964)]
[New Thread 0x7fffe544a700 (LWP 7965)]
[New Thread 0x7fffe2c49700 (LWP 7966)]
[New Thread 0x7fffe0448700 (LWP 7967)]
[New Thread 0x7fffddc47700 (LWP 7968)]
p>>> import tensorflow

Thread 1 "python" received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
(gdb) 
```

So seemingly things are okay _until_  the pthread_once machinery to initialize the thread pool is triggered. 

I'm a bit at a loss here on this, folks. I've spent about 4 hours now working on this and I don't know enough about what's going on inside to solve this. 

One solution we can try is to get rid of the global thread pool, and instead explicitly pass a thread pool into functions where one is supported. This means we could set up a thread pool that we can use for the lifetime of a pyarrow session and manage things in the Python bindings. This will definitely get rid of the issue. 

Another option is that we figure out where the thread pool is being initialized at import time and fix that so all of the imports, including tensorflow, can succeed. The downside is that if the user imports pyarrow, does something that uses threads, then later imports tensorflow, they will segfault.

**Comment by wesm** [comment]
FWIW, I'm +1 on offering more public thread control on functions that support it. In some applications you could imagine having separate thread pools for IO and computation

**Comment by robertnishihara** [comment]
Getting rid of the global thread pool may be the simplest option.

Unfortunately I think *"importing pyarrow, doing something that uses threads, then later importing tensorflow"* will be pretty common.

**Comment by pitrou** [comment]
Can someone try the workaround I mentioned in https://github.com/apache/arrow/pull/2198#issuecomment-401632426 ?

**Comment by pitrou** [comment]
By the way, I'm a bit surprised that `import pyarrow` would initialize the global thread pool. Do we know on which code path that happens?

**Comment by wesm** [comment]
@pitrou I will try the work around and also try to figure out where the thread pool is getting initialized. After thinking about this overnight, I think we should manage the global thread pool on the Python side. Then we are not relying on non-trivial global variables in the shared library

**Comment by pitrou** [comment]
The global thread pool is intended for C++ use as well. The prior situation was that we could launch local thread pools in several places (e.g. Plasma or memcpy) without regard for machine occupancy.

**Comment by wesm** [comment]
Right, I understand that. What I'm proposing is to expose `ThreadPool*` as a parameter instead of a flag `bool use_threads` anywhere where we have the global pool now. So the application can decide whether or not to use threads

**Comment by wesm** [comment]
I agree that this does expose some complexity to the developer-user; I see both sides of the argument, but I think it may be a good thing to offer more control/flexibility in the API in this way

**Comment by pitrou** [comment]
I don't know, I would tend to wait for user requests to come up :-)

**Comment by wesm** [comment]
OK, so I found out the mystery of the threads being spawned. They're being spawned by openblas when NumPy imports:

```
(gdb) break pthread_create
Function "pthread_create" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (pthread_create) pending.
(gdb) r
...
>>> import pyarrow
#0  __pthread_create_2_1 (newthread=0x7fffeedcd280 <blas_threads>, attr=0x0, 
    start_routine=0x7fffecbc08e0 <blas_thread_server>, arg=0x0) at pthread_create.c:610
#1  0x00007fffecbc0c08 in blas_thread_init ()
   from /home/wesm/miniconda/envs/wheel-test-3/lib/python3.6/site-packages/numpy/core/../.libs/libopenblasp-r0-39a31c03.2.18.so
#2  0x00007fffec9a0085 in gotoblas_init ()
   from /home/wesm/miniconda/envs/wheel-test-3/lib/python3.6/site-packages/numpy/core/../.libs/libopenblasp-r0-39a31c03.2.18.so
#3  0x00007ffff7de5733 in call_init (env=0x962b20, argv=0x7fffffffc4b8, argc=1, l=<optimized out>) at dl-init.c:72
#4  _dl_init (main_map=main_map@entry=0x963080, argc=1, argv=0x7fffffffc4b8, env=0x962b20) at dl-init.c:119
#5  0x00007ffff7dea1ff in dl_open_worker (a=a@entry=0x7fffffff1310) at dl-open.c:522
#6  0x00007ffff6a422df in __GI__dl_catch_exception (exception=0x7fffffff12f0, 
    operate=0x7ffff7de9dc0 <dl_open_worker>, args=0x7fffffff1310) at dl-error-skeleton.c:196
#7  0x00007ffff7de97ca in _dl_open (
    file=0x7fffef759b90 "/home/wesm/miniconda/envs/wheel-test-3/lib/python3.6/site-packages/numpy/core/multiarray.cpython-36m-x86_64-linux-gnu.so", mode=-2147483646, caller_dlopen=0x7ffff7a51161 <_PyImport_FindSharedFuncptr+417>, 
    nsid=<optimized out>, argc=1, argv=<optimized out>, env=0x962b20) at dl-open.c:605
#8  0x00007ffff7475f96 in dlopen_doit (a=a@entry=0x7fffffff1540) at dlopen.c:66
#9  0x00007ffff6a422df in __GI__dl_catch_exception (exception=exception@entry=0x7fffffff14e0, 
    operate=0x7ffff7475f40 <dlopen_doit>, args=0x7fffffff1540) at dl-error-skeleton.c:196
#10 0x00007ffff6a4236f in __GI__dl_catch_error (objname=0x69a350, errstring=0x69a358, mallocedp=0x69a348, 
    operate=<optimized out>, args=<optimized out>) at dl-error-skeleton.c:215
#11 0x00007ffff7476735 in _dlerror_run (operate=operate@entry=0x7ffff7475f40 <dlopen_doit>, 
    args=args@entry=0x7fffffff1540) at dlerror.c:162
#12 0x00007ffff7476051 in __dlopen (
<SNIP>
```

@pitrou I tried your workaround and it didn't work unfortunately. I'm going to take a crack at adding options to pass in a thread pool in the various places where we use it so we can unblock ourselves on this; not sure what else to do

**Comment by pitrou** [comment]
Changing our API just so that TensorFlow's non-compliant wheels don't crash sounds a bit over the board to me. I'd rather redirect user complaints to the TensorFlow project. We shouldn't feel responsible for this issue.

**Comment by robertnishihara** [comment]
Thanks @wesm! @pitrou I agree it'd be best to fix the issue on the TF side, but that's most likely a long-termer item. Not supporting TF at this point would be like being incompatible with numpy or pandas.

**Comment by wesm** [comment]
Here's the backtrace

```
#0  0x0000000000000000 in ?? ()
#1  0x00007ffff7688827 in __pthread_once_slow (
    once_control=0x7fffd4cd51c8 <tensorflow::port::(anonymous namespace)::cpuid_once_flag>, 
    init_routine=0x7ffff088bfe1 <std::__once_proxy()>) at pthread_once.c:116
#2  0x00007fffd4575faa in void std::call_once<void (&)()>(std::once_flag&, void (&)()) ()
   from /home/wesm/miniconda/envs/wheel-test-3/lib/python3.6/site-packages/tensorflow/python/../libtensorflow_framework.so
#3  0x00007fffd4575fde in tensorflow::port::TestCPUFeature(tensorflow::port::CPUFeature) ()
   from /home/wesm/miniconda/envs/wheel-test-3/lib/python3.6/site-packages/tensorflow/python/../libtensorflow_framework.so
#4  0x00007fffd4575f11 in tensorflow::port::(anonymous namespace)::CheckFeatureOrDie(tensorflow::port::CPUFeature, std::string const&) ()
   from /home/wesm/miniconda/envs/wheel-test-3/lib/python3.6/site-packages/tensorflow/python/../libtensorflow_framework.so
#5  0x00007fffd43b7394 in _GLOBAL__sub_I_cpu_feature_guard.cc ()
   from /home/wesm/miniconda/envs/wheel-test-3/lib/python3.6/site-packages/tensorflow/python/../libtensorflow_framework.so
#6  0x00007ffff7de5733 in call_init (env=0x944d20, argv=0x7fffffffc4a8, argc=1, l=<optimized out>) at dl-init.c:72
#7  _dl_init (main_map=main_map@entry=0xc29300, argc=1, argv=0x7fffffffc4a8, env=0x944d20) at dl-init.c:119
#8  0x00007ffff7dea1ff in dl_open_worker (a=a@entry=0x7fffffff57b0) at dl-open.c:522
#9  0x00007ffff6a422df in __GI__dl_catch_exception (exception=0x7fffffff5790, 
    operate=0x7ffff7de9dc0 <dl_open_worker>, args=0x7fffffff57b0) at dl-error-skeleton.c:196
#10 0x00007ffff7de97ca in _dl_open (
    file=0x7fffd9bcf640 "/home/wesm/miniconda/envs/wheel-test-3/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so", mode=-2147483646, caller_dlopen=0x7ffff7a51161 <_PyImport_FindSharedFuncptr+417>, 
    nsid=<optimized out>, argc=1, argv=<optimized out>, env=0x944d20) at dl-open.c:605
```

I'm pushing the branch as it stands now. I'll try a few more things

**Comment by pcmoritz** [comment]
I agree that fixing this in **some** way is extremely important. I can investigate the solution to try to import tensorflow before pyarrow if it is available. Any thoughts?

**Comment by wesm** [comment]
This is the cheapest thing I can think of:

```
$ python
Python 3.6.5 | packaged by conda-forge | (default, Apr  6 2018, 13:39:56) 
[GCC 4.8.2 20140120 (Red Hat 4.8.2-15)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ctypes
>>> import tensorflow
>>> tensorflow
<module 'tensorflow' from '/home/wesm/miniconda/envs/wheel-test-3/lib/python3.6/site-packages/tensorflow/__init__.py'>
>>> ctypes.CDLL('/home/wesm/miniconda/envs/wheel-test-3/lib/python3.6/site-packages/tensorflow/libtensorflow_framework.so')
<CDLL '/home/wesm/miniconda/envs/wheel-test-3/lib/python3.6/site-packages/tensorflow/libtensorflow_framework.so', handle 23d7f40 at 0x7f3834cc6080>
>>> import pyarrow
>>> import tensorflow
WORKS
```

Loading the tensorflow shared library is fast, whereas `import tensorflow` takes about as long as importing numpy

**Comment by pcmoritz** [comment]
Yep, that's the way to do it. Let me see if I can get the path without importing tensorflow ;)

**Comment by wesm** [comment]
Probably have to look in site-packages

```
>>> import site
>>> site.getsitepackages()
['/home/wesm/miniconda/envs/wheel-test-3/lib/python3.6/site-packages']
```

suffice to say this workaround should only be needed on Linux

**Comment by wesm** [comment]
OK @pcmoritz I will leave this branch to you.

@pitrou I changed the various ThreadPool variable to use `unique_ptr` instead of `shared_ptr`; I can change it back, but I wonder if you feel strongly about one vs. the other?

**Comment by wesm** [comment]
Roger, reverted

**Comment by robertnishihara** [comment]
Closing because hopefully this is worked around by #2210.

**Pull request closed without merging**